### PR TITLE
change event hub package to GA

### PR DESCRIPTION
**Pull request by Luyunmt**
change event hub package to GA

### CODE REVIEW

**Comment by ramya-rao-a** [comment]
@jsquire, Can you please review this PR?

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedFunction.csproj:7]: Please bump to the latest GA,  v5.2.0

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedFunction.csproj:3]: If there a reason why we would use a down-level framework?   I'd recommend bumping to `netcoreapp3.1` and targeting Functions v3 for the latest LTS.  If we're set on targeting Functions 2, then we should be using `netcoreapp2.1`.    We shouldn't be targeting Functions v1 in a sample.   
_([Azure Functions supported versions](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library?tabs=v2%2Ccmd#supported-versions))_

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedFunction.csproj:4]: Assuming the target framework is bumped, this can be changed to `latest`.    We should try to avoid `preview` for samples.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:39]: This should return `Task`.    In general, no method should be returning `void` if marked `async` outside of a very few special corner cases.   Here, Functions supports the correct signature.   _([ref](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library?tabs=v2%2Ccmd#async))_

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:52]: We should not be demonstrating obsolete members in a sample.   Any idea what the compiler is flagging here?

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:55]: This isn't the correct pattern to use.   The producer client should be created as a `static` member of the class, following the pattern shown in the `DocumentClient` sippet, [here](https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections#documentclient-code-example-c).

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:54]: nit:  The product name is `Event Hubs` of which there is one ore more `Event Hub` instances.   We should be using this terminology and respecting capitalization.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:51]: nit:  The product name is `Event Hubs` of which there is one ore more `Event Hub` instances.   We should be using this terminology and respecting capitalization.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:66]: nit:  The product name is `Event Hubs` of which there is one ore more `Event Hub` instances.   We should be using this terminology and respecting capitalization.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:64]: You're not checking the return value of `TryAdd` which can lead to data loss.   In the scenario where the batch is full, the loop will continue to iterate and events will silently be ignored.    Please either follow the pattern to check/throw _([ref](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample04_PublishingEvents.md#publishing-size-constraints))_ or  demonstrate how to build multiple batches to accommodate all documents, as illustrated [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample04_PublishingEvents.md#creating-and-publishing-multiple-batches).

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:61]: nit:  JSON

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:36]: Line 31 should be `const` rather than `static readonly`

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:62]: We should update this to use `System.Text.Json` and remove the `Newtonsoft` reference.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:66]:
```suggestion
                // Use the producer to send the change events to Event Hubs.

```

**Review by jsquire — CHANGES_REQUESTED** [review]

**Comment by jsquire** [comment]
The `obj` folder under the `DataGenerator` project should not be in source control.

**Comment by v-xuto** [comment]
@jsquire, We have updated, can you please re-review this PR?

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedFunction.csproj:7]: Can we please bump this to 5.4.1?

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:71]: This is going to spew the entire JSON blob back in the error message.   Is there a way that we can distill that into something more readable?   Maybe reporting the index into the collection where things failed?

**Review by jsquire — APPROVED** [review]

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs:71]: I'm investigating. It may take a while.

Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: It has been updated.

**Comment by v-xuto** [comment]
@jsquire - We have updated this PR, please re-review this PR.

**Comment by v-xuto** [comment]
@scottaddie, @AlexGhiondea, @lilyjma, @ramya-rao-a  Could you please review this PR?

**Comment by ramya-rao-a** [comment]
@jsquire Can you take another look here? The PR has been updated after your last approval

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs:71]: @jsquire - I have updated, please review.

**Comment by v-xuto** [comment]
@jsquire - We have updated this PR, please re-review this PR.

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedFunction.csproj:7]:
```suggestion
    <PackageReference Include="Azure.Messaging.EventHubs" Version="5.6.1" />

```

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:36]: Is there a reason that we're not using [dependency injection](https://docs.microsoft.com/azure/azure-functions/functions-dotnet-dependency-injection) to configure the producer.   That would be the recommended approach.  In `ConfigureServices`, the `AddEventHubProducerClient` extension can be used to register the producer as a singleton and ensure proper clean-up.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:72]:
```suggestion
                EventData data = new EventData(json);
```

No need to encode; there's an overload that handles this now.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs:71]: I still have the came concerns.   If we're attempting to demonstrate recommended patterns for customer applications, we'd want to handle the case where we need to break the set into multiple batches.   A simple illustration can be found [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample04_PublishingEvents.md#creating-and-publishing-multiple-batches).

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: @jsquire - I have updated, please review.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: @jsquire - I have updated, please review.

File: ChangeFeedLabSolution/ChangeFeedFunction.csproj
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedFunction.csproj]: @jsquire - I have updated, please review.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: @jsquire - I have updated, please review.

**Comment by v-xuto** [comment]
@jsquire - We have updated, please rereview this PR.

**Comment by v-xuto** [comment]
@jsquire - We have updated, please rereview this PR.

**Comment by v-xuto** [comment]
@jaschrep-msft - We have updated by your comments, please re-review this PR.

**Comment by v-xuto** [comment]
@jsquire - We have updated, please rereview this PR.

**Comment by v-xuto** [comment]
@jsquire - We have updated by your comments, please rereview this PR.

**Comment by ramya-rao-a** [comment]
@v-xuto This is the holiday season and you may get low response. Please circle back in January

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:79]: Have we actually run this code?

The producer is registered with DI as a singleton;  if you close it here, it can no longer be used and subsequent invocations of the function will fail.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:32]: This is a really unusual format.   A common C# convention is that class-level members are prefixed with an underscore to denote their scope.    If we're going to use prefixes, then we should be doing something similar - I've never seen parameters decorated like this.

Alternatively, we could make the member  `private readonly EventHubProducerClient EventHubProducerClient;` to disambiguate.   Either is fine, as would keeping the same private member name and disambiguating with `this`.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:28]: This comment does not match the member that it is associated with.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (jsquire) [inline ChangeFeedLabSolution/ChangeFeedProcessor.cs:82]:
```suggestion
            private static async Task<IReadOnlyList<EventDataBatch>> BuildBatchesAsync(

```

File: ChangeFeedLabSolution/Startup.cs
Comment (jsquire) [inline ChangeFeedLabSolution/Startup.cs:10]: This should read from configuration rather than being hardcoded here.

File: ChangeFeedLabSolution/Startup.cs
Comment (jsquire) [inline ChangeFeedLabSolution/Startup.cs:17]: why are we creating a producer here and then doing nothing with it?   I think we can simplify this to just use the `AddEventHubProducerClient` call.

File: ChangeFeedLabSolution/local.settings.json
Comment (jsquire) [inline ChangeFeedLabSolution/local.settings.json:10]: By [spec](https://www.json.org/json-en.html), JSON does not support embedded comments, even though some parsers tolerate them.   We should stick to standard JSON and remove the comments.

_([additional context](https://web.archive.org/web/20150105080225/https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr))_

**Review by jsquire — CHANGES_REQUESTED** [review]

**Comment by jsquire** [comment]
I'm concerned that we're not running these samples after making changes.   Can we please be sure to add a validation step for running the code before submitting the PR or requesting a re-review?

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: @jsquire - Updated.

File: ChangeFeedLabSolution/Startup.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/Startup.cs]: @jsquire - Removed.

File: ChangeFeedLabSolution/local.settings.json
Comment (Tianyun-G) [reply ChangeFeedLabSolution/local.settings.json]: @jsquire - Removed.

File: ChangeFeedLabSolution/Startup.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/Startup.cs]: @jsquire - Updated.

File: ChangeFeedLabSolution/ChangeFeedProcessor.cs
Comment (Tianyun-G) [reply ChangeFeedLabSolution/ChangeFeedProcessor.cs]: @jsquire - Removed.

**Comment by Tianyun-G** [comment]
@jsquire - I have updated and running these samples, please review.

File: ChangeFeedLabSolution/Startup.cs
Comment (jsquire) [inline ChangeFeedLabSolution/Startup.cs:14]: What happens if we want to read these values from `local.settings.json` where they're also specified?   We should be using the Functions configuration types.   Please see:  https://docs.microsoft.com/azure/azure-functions/functions-dotnet-dependency-injection#working-with-options-and-settings

Comment (Tianyun-G) [reply ChangeFeedLabSolution/Startup.cs:14]: @jsquire -  We can also read these values from `local.settings.json` using `Environment.GetEnvironmentVariable("")`

**Comment by Tianyun-G** [comment]
@jsquire - I have updated as your recommend ,please review.

**Review by jsquire — APPROVED** [review]

**Comment by jsquire** [comment]
@jongio:  The conversion is complete.   Would you be so kind as to help get this merged?

**Pull request merged**
